### PR TITLE
OKTA-419312: Improve keyboard focus handling in odyssey-react Modal component via @react-aria/focus

### PR DESCRIPTION
**Pull request by kamalgill-okta**
### Description

This PR introduces the following features for the odyssey-react Modal component...

- When the modal is OPEN the first focusable item should receive focus (in the case of the modal it is the Dismiss button)
- The focus should be retained within the bounds of the modal dialog at all times while the modal is in view.
- When the modal is CLOSED the element focused prior to the modal opening should receive focus.

See https://oktainc.atlassian.net/browse/OKTA-419312 (Okta employees only)

NOTE: This PR replaces https://github.com/okta/odyssey/pull/1362

### CODE REVIEW

File: packages/odyssey-react/src/utils/focusHandling.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  setFocus: (elem: OptionalHTMLElement) => OptionalHTMLElement;
+}
+
+const FOCUSABLE_ITEMS = [
+  "button",
+  "[href]",
+  "input",
+  "select",
+  "textarea",
+  '[tabindex]:not([tabindex="-1"])',
```
Comment (jeffbelser-okta) [inline packages/odyssey-react/src/utils/focusHandling.ts:26]: bumping previous feedback about the IE11 support for the `not` selector, it doesn't have it! We'll need to adjust the query string here and then separately inspect the value of this query to filter out tab indicies below 0.

File: packages/odyssey-react/src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -111,9 +122,14 @@ export const Modal = withTheme(
...
-    if (open && onOpen) {
-      onOpen();
+    const { restoreFocus, setFocus } = useFocus();
+    const lastFocusedElemRef = useRef<OptionalHTMLElement>(null);
+
+    if (open) {
+      lastFocusedElemRef.current = setFocus(modalDialog.current);
+      onOpen && onOpen();
+    } else {
+      lastFocusedElemRef.current && restoreFocus(lastFocusedElemRef.current);
```
Comment (jeffbelser-okta) [inline packages/odyssey-react/src/components/Modal/Modal.tsx:132]: the issue with shadowed props and state are no longer problematic? I can close the modal and re-open it now without the bug where the prop stays closed?

File: packages/odyssey-react/src/components/Modal/Modal.test.tsx [hunk 3]
```diff
@@ -130,6 +130,18 @@ describe("Modal", () => {
...
   });
 
+  xit("should automatically focus on the dismiss icon when modal is open", () => {
+    const handleClose = jest.fn();
+    render(
+      <Modal open onClose={handleClose} closeMessage={message}>
+        <Modal.Header>{modalHeading}</Modal.Header>
+      </Modal>
+    );
+    const dismissIcon = screen.getByTitle(message).closest("button");
```
[the remark this answers is not in the document]
Comment (kamalgill-okta) [reply packages/odyssey-react/src/components/Modal/Modal.test.tsx:145]: The tests now pass, so I think we're fine here.

File: packages/odyssey-react/src/utils/focusHandling.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  );
+  // Capture original focused element before setting focus inside modal dialog
+  const lastFocusedElement = document.activeElement;
+  if (focusableItems.length > 0) {
+    requestAnimationFrame(() => {
+      // Focus on first focusable element inside dialog
+      focusableItems[0].focus();
+    });
+  }
+  return lastFocusedElement as OptionalHTMLElement;
```
[the remark this answers is not in the document]
Comment (kamalgill-okta) [reply packages/odyssey-react/src/utils/focusHandling.ts:51]: I'm not sure I understand what you're suggesting w.r.t. generics here -- feel free to refactor if needed.

File: packages/odyssey-react/src/components/Modal/Modal.test.tsx [hunk 5]
```diff
@@ -130,6 +130,52 @@ describe("Modal", () => {
...
+        <button data-testid="modal-trigger">Open Modal</button>
+        <Modal open={false} onClose={handleClose} closeMessage={message}>
+          <Modal.Header>{modalHeading}</Modal.Header>
+        </Modal>
+      </div>
+    );
+    const triggerBtn = screen.getByTestId("modal-trigger");
+    triggerBtn && triggerBtn.click();
+    const dismissIcon = screen.getByTitle(message).closest("button");
+    waitFor(
```
[the remark this answers is not in the document]
Comment (kamalgill-okta) [reply packages/odyssey-react/src/components/Modal/Modal.test.tsx:146]: bummer -- I probably misunderstood how `waitFor` works, and adding async/await to the tests aren't allowing them to pass.  I'll probably need to hand this over to the Odyssey team devs to sort this out given my obvious lack of knowledge around React Testing Library and how it handles (or doesn't handle) `window.requestAnimationFrame()`.

File: packages/odyssey-react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (kamalgill-okta) [reply packages/odyssey-react/src/components/Modal/Modal.tsx:132]: Correct, it handles multiple open/close cycles since we're not managing the open/close state in the component.  Incidentally, given the Modal component doesn't manage open/closed state, it's quite difficult (if not impossible) to have the component internally manage the open/closed state, preventing us from implementing the "close on `Esc` key press" and "close when clicking outside the modal" behaviors without a significant amount of work.  Those efforts are tracked in a separate ticket and are outside the scope of this PR.

File: packages/odyssey-react/src/components/Modal/Modal.test.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeffbelser-okta) [reply packages/odyssey-react/src/components/Modal/Modal.test.tsx:146]: no this is a great start, I think you have 90% of the moving parts here just trying to protect against some test edge cases that lead to flake for async actions!

File: packages/odyssey-react/src/utils/focusHandling.ts (hunk 1 above)
Comment (andrewberg-okta) [reply packages/odyssey-react/src/utils/focusHandling.ts]: Addressed in 6943007

File: packages/odyssey-react/src/utils/focusHandling.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffbelser-okta) [reply packages/odyssey-react/src/utils/focusHandling.ts:57]: @kamalgill-okta I think we can improve the type safety here by using a generic type function. Yes something to refactor

File: packages/odyssey-react/src/components/Modal/Modal.tsx (hunk 2 above)
Comment (jeffbelser-okta) [reply packages/odyssey-react/src/components/Modal/Modal.tsx:132]: @kamalgill-okta thanks these are all problems we can solve!
